File: packages/ws-bridge/src/frame.ts
// Frame types matching the Node ↔ Python bridge protocol.
// See docs/node-python-bridge.md §2 for the wire format.

export type FrameKind = 'request' | 'response' | 'event' | 'error' | 'ping' | 'pong' | 'hello';

export interface BridgeFrame {
  v: number; // protocol version, == 1
  id: string; // ULID
  kind: FrameKind;
  ts?: number; // unix ms
  method?: string; // required for request / event
  params?: Record<string, unknown>;
  result?: Record<string, unknown>;
  error?: {
    code: string;
    message: string;
    data?: Record<string, unknown>;
    retryable?: boolean;
  };
  auth?: string; // hello frame only
  corr?: string; // correlation id
}

// ── Builders ──

let _ulidCounter = 0;

/** Generate a simple ULID-like id (26 chars, timestamp-prefixed). */
export function ulid(): string {
  const ts = Date.now().toString(32).padStart(10, '0');
  const rand = Math.random().toString(32).slice(2, 12);
  const count = (++_ulidCounter % 10000).toString(32).padStart(4, '0');
  return `${ts}${rand}${count}`;
}

export function helloFrame(auth: string, client = 'cli', ver = '0.1.0'): BridgeFrame {
  return {
    v: 1,
    id: ulid(),
    kind: 'hello',
    ts: Date.now(),
    auth,
    params: { client, ver },
  };
}

export function requestFrame(method: string, params?: Record<string, unknown>): BridgeFrame {
  const frame: BridgeFrame = {
    v: 1,
    id: ulid(),
    kind: 'request',
    ts: Date.now(),
    method,
  };
  if (params !== undefined) frame.params = params;
  return frame;
}

export function pingFrame(): BridgeFrame {
  return {
    v: 1,
    id: ulid(),
    kind: 'ping',
    ts: Date.now(),
  };
}

export function cancelFrame(targetId: string): BridgeFrame {
  return requestFrame('task.cancel', { target: targetId });
}

File: backend/app/gateway/frame.py
"""Gateway frame models — Pydantic schemas matching the Node ↔ Python bridge protocol.

See docs/node-python-bridge.md §2 for the wire format.
"""

from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


# --- Envelope kinds ---
FrameKind = Literal["request", "response", "event", "error", "ping", "pong", "hello"]


class BridgeFrame(BaseModel):
    """Top-level envelope for all ws frames."""

    v: int = Field(default=1, description="Protocol version")
    id: str = Field(description="ULID, client-generated per request; server-generated for events")
    kind: FrameKind = Field(description="Frame kind")
    ts: int = Field(default=0, description="Unix ms timestamp")
    method: Optional[str] = Field(default=None, description="Method name (required for request/event)")
    params: Optional[dict[str, Any]] = Field(default=None, description="Request/event payload")
    result: Optional[dict[str, Any]] = Field(default=None, description="Response payload")
    error: Optional[dict[str, Any]] = Field(default=None, description="Error payload")
    auth: Optional[str] = Field(default=None, description="Bearer token (hello frame only)")
    corr: Optional[str] = Field(default=None, description="Correlation id, linking response to request")

    # ── convenience builders ──

    @classmethod
    def response(cls, *, corr: str, result: Optional[dict[str, Any]] = None) -> BridgeFrame:
        """Build a response frame."""
        return cls(
            kind="response",
            id=corr,
            corr=corr,
            result=result or {},
        )

    @classmethod
    def error_response(
        cls,
        *,
        corr: str,
        code: str,
        message: str,
        data: Optional[dict[str, Any]] = None,
        retryable: bool = False,
    ) -> BridgeFrame:
        """Build an error response frame."""
        return cls(
            kind="error",
            id=corr,
            corr=corr,
            error={
                "code": code,
                "message": message,
                "data": data or {},
                "retryable": retryable,
            },
        )

    @classmethod
    def event(cls, *, method: str, params: dict[str, Any]) -> BridgeFrame:
        """Build a server-pushed event frame."""
        import ulid

        return cls(
            kind="event",
            id=str(ulid.ULID()),
            method=method,
            params=params,
        )

    @classmethod
    def pong(cls, *, corr: str) -> BridgeFrame:
        """Build a pong response to a ping."""
        return cls(kind="pong", id=corr, corr=corr)
